### DISCUSSION TITLE
The objective is not DCP - QuadraticFunction problem

### DISCUSSION THREAD

**Discussion by mrbooOoo** [cvxpy/cvxpy#1468] [Q&A]
Dear CVXPY folks,

I've been using cvxpy the last 1/2 month(s) and it has been great. Today I was trying to implement a quadratic function in the objective function but without success. 

````python
# --- Agents ---
m = 4
n = 4
# --------  Data  ---------
Pmax = np.array([100,50,100,80])
price = np.array([40,35,12,20])
agent_sign = np.array([1,1,-1,-1])
bilateral_matrix = np.array([[0,0,1,1],[0,0,1,1],[1,1,0,0],[1,1,0,0]])
# -----------    Parameters     -------------
Pmax_n = cp.Parameter(n, value=Pmax)
price_n = cp.Parameter(n, value=price)
agent_sign_n = cp.Parameter(n, value=agent_sign)
bilateral_matrix_nm = cp.Parameter((n,m), value=bilateral_matrix)
#  -----   Variables   -------
Pn = cp.Variable(n, nonneg=True)
Pnm = cp.Variable((n,m), nonneg=True)
# ------ -- Objective Function ---------
obj_function = cp.sum(cp.multiply(cp.power(Pn,2),cp.multiply(agent_sign_n,price_n)))
SW_problem = cp.Maximize(obj_function)
````

This is a simple example of what i'm trying to implement. 
Basically i want to maximize z and <img src="https://render.githubusercontent.com/render/math?math=z=\sum_n( P_{n}^{2}*price_n*agentsign_n)">

This generates an error : raise DCPError(
cvxpy.error.DCPError: Problem does not follow DCP rules. Specifically:
The objective is not DCP. Its following subexpressions are not:
power(var4, 2.0) @ param2 @ param1

Can someone help me? 

I will attatch the file.
[ex1-Full-P2P-cvxpy.zip](https://github.com/cvxpy/cvxpy/files/7068752/ex1-Full-P2P-cvxpy.zip)

**Comment by SteveDiamond** [comment]
This doesn't look like a convex problem to me. You're summing a mix of positive and negative quadratics. Perhaps you can write the objective as a single quadratic: ``quad_form(Pn, P)``. Here P is a positive semidefinite matrix.

**Reply by mrbooOoo** [reply]
Thank you for you help, but that won't solve my problem. Because i really need to square Pn (which is a variable).

**Reply by mrbooOoo** [reply]
I was doing some search and i represented the function, and it is convex. So, basically this is a convex problem, it just have one possible solution.

**Reply by SteveDiamond** [reply]
The example above is not convex. You can use this tool to prove that it is not convex: https://github.com/cvxgrp/cvxpyanalyzer

**Reply by mrbooOoo** [reply]
Yes, you are right it is a concave problem. There's any way i can solve this, this way? 
If I change the signal of the objective function, the problem changes from concave to convex? 

I saw you guys developed a library for convex problems, using PyTorch or TensorFlow, but I don't know if that were a solution at all...

Thank you for your help and understanding, Steve. I appreciate it.

**Reply by SteveDiamond** [reply]
Sorry, that was confusing. When I said the problem is not convex I meant the objective is not concave. Again I'd suggest trying the analyzer tool. Are you sure you're inputting the problem correctly to gurobi? I would be surprised if gurobi can handle it but cvxpy can't.

**Reply by mrbooOoo** [reply]
Hey Steve, i tried the analyzer tool. When I run some of the functions, it gives me errors like this one:

Latest CVXPY version: 1.1.15
Local CVXPY version: 1.1.15
You have the latest version of CVXPY.
The problem is not DCP. You must write the problem so it follows the DCP rules.
Learn more about the DCP rules at https://dcp.stanford.edu
Traceback (most recent call last):
  File "c:\Users\cj\OneDrive - Universidade de Tras-os-Montes e Alto Douro\INESC TEC\DECARBONIZE\IEEE-PES-conference\P2P-market\Full-P2P-benchmark.py", line 218, in <module>
    tech_support(problem)
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\analyzer\tech_support.py", line 61, in tech_support
    convexish = might_be_convex(problem)
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\analyzer\convexity_checker.py", line 49, in might_be_convex
    f1 = get_objective_constraint_values(problem, maximize)
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\analyzer\convexity_checker.py", line 39, in get_objective_constraint_values
    + [np.atleast_1d(c.expr.value) for c in problem.constraints]
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\analyzer\convexity_checker.py", line 39, in <listcomp>
    + [np.atleast_1d(c.expr.value) for c in problem.constraints]
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\cvxpy\atoms\atom.py", line 349, in value
    return self._value_impl()
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\cvxpy\atoms\atom.py", line 363, in _value_impl
    arg_val = arg._value_impl()
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\cvxpy\atoms\atom.py", line 368, in _value_impl
    result = self.numeric(arg_values)
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\cvxpy\atoms\affine\sum.py", line 60, in numeric
    result = np.sum(values[0], axis=self.axis, keepdims=self.keepdims)
  File "<__array_function__ internals>", line 5, in sum
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\numpy\core\fromnumeric.py", line 2259, in sum
    return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims,
  File "C:\Users\cj\Anaconda3\envs\cvxpy\lib\site-packages\numpy\core\fromnumeric.py", line 86, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
numpy.AxisError: axis 1 is out of bounds for array of dimension 1

**Reply by SteveDiamond** [reply]
Ok I did some research,  and actually Gurobi can handle non-convex quadratic objectives (https://www.gurobi.com/resource/non-convex-quadratic-optimization/).

You might want to look into QCQP, which is a CVXPY extension for non-convex quadratic problems (https://github.com/cvxgrp/qcqp). It's only compatible with cvxpy 0.4.9 for the moment.

**Reply by mrbooOoo** [reply]
Ok, i will have a look and try to solve my problem with it. Thanks, Steve. Thanks for your understanding!
